File: frontend/src/services/customerService.ts
import api from '@/lib/api'

export interface Customer {
  id: number
  organization_id: number
  name: string
  email: string
  phone: string
  address: string
  created_at: string
  updated_at: string
}

export interface CreateCustomerData {
  name: string
  email: string
  phone: string
  address: string
}

export interface PaginationResponse {
  page: number
  limit: number
  total: number
  total_pages: number
}

export interface CustomerListResponse {
  customers: Customer[]
  pagination: PaginationResponse
}

export const customerService = {
  async list(search?: string, page = 1, limit = 20): Promise<CustomerListResponse> {
    const params = new URLSearchParams()
    if (search) params.append('search', search)
    params.append('page', page.toString())
    params.append('limit', limit.toString())

    const response = await api.get(`/customers?${params}`)
    return {
      customers: response.data.data,
      pagination: response.data.pagination,
    }
  },

  async getById(id: number): Promise<Customer> {
    const response = await api.get(`/customers/${id}`)
    return response.data.data
  },

  async create(data: CreateCustomerData): Promise<Customer> {
    const response = await api.post('/customers', data)
    return response.data.data
  },

  async update(id: number, data: CreateCustomerData): Promise<Customer> {
    const response = await api.put(`/customers/${id}`, data)
    return response.data.data
  },

  async delete(id: number): Promise<void> {
    await api.delete(`/customers/${id}`)
  },
}

File: internal/service/customer_service.go
package service

import (
	"invoice-saas/internal/model"
	"invoice-saas/internal/repository"
	"invoice-saas/pkg/validator"
)

type CustomerService struct {
	repo *repository.CustomerRepository
}

func NewCustomerService(repo *repository.CustomerRepository) *CustomerService {
	return &CustomerService{repo: repo}
}

func (s *CustomerService) Create(organizationID int, req *model.CreateCustomerRequest) (*model.Customer, validator.ValidationErrors) {
	errors := validator.ValidationErrors{}

	if !validator.ValidateRequired(req.Name) {
		errors.Add("name", "Name is required")
	}

	if req.Email != "" && !validator.ValidateEmail(req.Email) {
		errors.Add("email", "Invalid email address")
	}

	if errors.HasErrors() {
		return nil, errors
	}

	customer := &model.Customer{
		OrganizationID: organizationID,
		Name:           req.Name,
		Email:          req.Email,
		Phone:          req.Phone,
		Address:        req.Address,
	}

	if err := s.repo.Create(customer); err != nil {
		errors.Add("name", "Failed to create customer")
		return nil, errors
	}

	return customer, nil
}

func (s *CustomerService) Update(id, organizationID int, req *model.UpdateCustomerRequest) (*model.Customer, validator.ValidationErrors) {
	errors := validator.ValidationErrors{}

	if !validator.ValidateRequired(req.Name) {
		errors.Add("name", "Name is required")
	}

	if req.Email != "" && !validator.ValidateEmail(req.Email) {
		errors.Add("email", "Invalid email address")
	}

	if errors.HasErrors() {
		return nil, errors
	}

	customer := &model.Customer{
		ID:             id,
		OrganizationID: organizationID,
		Name:           req.Name,
		Email:          req.Email,
		Phone:          req.Phone,
		Address:        req.Address,
	}

	if err := s.repo.Update(customer); err != nil {
		errors.Add("name", "Failed to update customer")
		return nil, errors
	}

	updated, err := s.repo.FindByID(id, organizationID)
	if err != nil {
		errors.Add("customer", "Failed to retrieve updated customer")
		return nil, errors
	}

	return updated, nil
}

func (s *CustomerService) Delete(id, organizationID int) error {
	return s.repo.Delete(id, organizationID)
}

func (s *CustomerService) GetByID(id, organizationID int) (*model.Customer, error) {
	return s.repo.FindByID(id, organizationID)
}

func (s *CustomerService) List(organizationID int, search string, page, limit int) ([]model.Customer, *model.PaginationResponse, error) {
	if page < 1 {
		page = 1
	}
	if limit < 1 || limit > 100 {
		limit = 20
	}

	customers, total, err := s.repo.List(organizationID, search, page, limit)
	if err != nil {
		return nil, nil, err
	}

	totalPages := (total + limit - 1) / limit

	pagination := &model.PaginationResponse{
		Page:       page,
		Limit:      limit,
		Total:      total,
		TotalPages: totalPages,
	}

	return customers, pagination, nil
}
